internal/scheduler/legacyscheduler.go

```diff
@@ -318,151 +400,180 @@ func uuidFromUlidString(ulid string) (uuid.UUID, error) {
-		usageAtPriority, ok := usageByPriority[priority]
-		if ok {
-			total.Add(usageAtPriority)
-		}
-		cumulativeUsageByPriority[priority] = total.DeepCopy()
+		cumulativeSum.Add(usedByPriority[priority])
+		cumulativeUsageByPriority[priority] = cumulativeSum.DeepCopy()
 	}
-
 	for priority, priorityLimits := range limits {
 		if priority <= jobPriority {
-			rl, ok := cumulativeUsageByPriority[priority]
-			if ok {
-				limitExceeded, msg := sched.exceedsResourceLimits(ctx, rl, priorityLimits)
+			if rl, ok := cumulativeUsageByPriority[priority]; ok {
+				limitExceeded, msg := exceedsResourceLimits(ctx, rl, total, priorityLimits)
 				if limitExceeded {
 					return true, fmt.Sprintf("%s at priority %d", msg, priority)
 				}
+			} else {
+				log.Warnf("Job scheduled at priority %d but there are no per-priority limits set up for this class; skipping per periority limit check", priority)
 			}
 		}
 	}
 	return false, ""
 }
 
 // Check that this job is at least equal to the minimum job size.
-func (sched *LegacyScheduler) jobIsLargeEnough(jobTotalResourceRequests common.ComputeResources) (bool, string) {
-	// TODO: These per-job checks could be expressed as filter functions, e.g., of type
-	// jobsFilterFunc func(*api.Job) bool
-	if len(sched.MinimumJobSize) == 0 {
+func jobIsLargeEnough(jobTotalResourceRequests, minimumJobSize schedulerobjects.ResourceList) (bool, string) {
+	if len(minimumJobSize.Resources) == 0 {
 		return true, ""
 	}
-	if len(jobTotalResourceRequests) == 0 {
+	if len(jobTotalResourceRequests.Resources) == 0 {
 		return true, ""
 	}
-	for resourceType, limit := range sched.MinimumJobSize {
-		if q, ok := jobTotalResourceRequests[resourceType]; ok {
-			if limit.Cmp(q) == 1 {
-				return false, fmt.Sprintf(
-					"job requests %s %s, but the minimum is %s",
-					q.String(), resourceType, limit.String(),
-				)
-			}
-		} else {
+	for resourceType, limit := range minimumJobSize.Resources {
+		q := jobTotalResourceRequests.Get(resourceType)
+		if limit.Cmp(q) == 1 {
 			return false, fmt.Sprintf(
-				"job requests 0 %s, but the minimum is %s",
-				resourceType, limit.String(),
+				"job requests %s %s, but the minimum is %s",
+				q.String(), resourceType, limit.String(),
 			)
 		}
 	}
 	return true, ""
 }
 
-func (sched *LegacyScheduler) selectNodeForPod(ctx context.Context, jobId uuid.UUID, job *api.Job, bind bool) (*PodSchedulingReport, error) {
-	podSpec := util.PodSpecFromJob(job)
-	if podSpec == nil {
-		return nil, errors.New("failed to get pod spec")
+type CandidateGangIterator struct {
+	ctx context.Context
+	CandidateJobsIterator
+	// Jobs are grouped into gangs by this annotation.
+	gangIdAnnotation string
+	// Jobs in a gang specify the number of jobs in the gang via this annotation.
+	gangCardinalityAnnotation string
+	// Map from queue to the gangs for that queue.
+	// Each gang is made up of a slice of jobs with equal value for gangIdAnnotation.
+	gangsByQueue map[string]map[string][]*JobSchedulingReport
```

Comment:
It looks like the `jobSpechedulingReport` seems to be moving from being a "Report" (i.e. a record of what happened), to a "context" (i.e. something that's storing the state of the entire scheduling round).  Is it worth renaming (or even refactoring) this.

Reply:
I agree. It may be best to do in a subsequent pr.